lib/android-helpers.js

```diff
@@ -372,32 +373,54 @@ helpers.pushStrings = async function (language, adb, opts) {
   return {};
 };
 
-helpers.unlock = async function (adb) {
+helpers.unlockWithUIAutomation = async function (driver, adb, capabilities) {
+  let unlockType = capabilities.unlockType;
+  if (!unlocker.isValidUnlockType(unlockType)) {
+    throw new Error(`Invalid unlock type ${unlockType}`);
+  }
+  let unlockKey = capabilities.unlockKey;
+  if (!unlocker.isValidKey(unlockType, unlockKey)) {
+    throw new Error(`Missing unlockKey ${unlockKey} capability for unlockType ${unlockType}`);
+  }
+  const unlockMethod = {
+    [PIN_UNLOCK]: unlocker.pinUnlock,
+    [PASSWORD_UNLOCK]: unlocker.passwordUnlock,
+    [PATTERN_UNLOCK]: unlocker.patternUnlock,
+  }[unlockType];
+  await unlockMethod(adb, driver, capabilities);
+};
+
+helpers.unlockWithHelperApp = async function(adb) {
+  logger.info("Unlocking screen");
+  await adb.forceStop('io.appium.unlock');
+  // then start the app twice, as once is flakey
+  let startOpts = {
+    pkg: "io.appium.unlock",
+    activity: ".Unlock",
+    action: "android.intent.action.MAIN",
+    category: "android.intent.category.LAUNCHER",
+    flags: "0x10200000",
+    stopApp: false
+  };
+  await adb.startApp(startOpts);
+  await adb.startApp(startOpts);
+};
+
+helpers.unlock = async function (driver, adb, capabilities) {
   if (!(await adb.isScreenLocked())) {
     logger.info("Screen already unlocked, doing nothing");
     return;
   }
-  logger.info("Unlocking screen");
-
-  await retryInterval(10, 1000, async () => {
+  await retryInterval(1, 1000, async () => {
     logger.debug("Screen is locked, trying to unlock");
-
-    // first manually stop the unlock activity
-    await adb.forceStop('io.appium.unlock');
-
-    // then start the app twice, as once is flakey
-    let startOpts = {
-      pkg: "io.appium.unlock",
-      activity: ".Unlock",
-      action: "android.intent.action.MAIN",
-      category: "android.intent.category.LAUNCHER",
-      flags: "0x10200000",
-      stopApp: false
-    };
-    await adb.startApp(startOpts);
-    await adb.startApp(startOpts);
-
     // check if it worked, twice
+    if (typeof capabilities.unlockType === 'undefined') {
+      // Leave the old unlock to avoid breaking existing tests
+      logger.warn("Using app unlock, this is going to be deprecated!");
```

Comment:
alternatively, just refactor so that the conditional is outside the retry:

```js
if (typeof capabilities.unlockType === 'undefined') {
  logger.warn("Using app unlock, this is going to be deprecated!");
  await retryInterval(10, 1000, async () => {
    await helpers.unlockWithHelperApp(adb);
  });
} else {
  await helpers.unlockWithUIAutomation(driver, adb, {unlockType: capabilities.unlockType, unlockKey: capabilities.unlockKey});
}
```